qiskit/assembler/assemble_circuits.py

```diff
@@ -105,21 +132,112 @@ def _assemble_circuit(circuit):
+        return None, None
+    if not hasattr(run_config, 'parametric_pulses'):
+        run_config.parametric_pulses = []
+    calibrations = []
+    pulse_library = {}
+    for gate, cals in circuit.calibrations.items():
+        for (qubits, params), schedule in cals.items():
+            qobj_instructions, _ = _assemble_schedule(
+                schedule,
+                converters.InstructionToQobjConverter(PulseQobjInstruction),
+                run_config,
+                pulse_library)
+            calibrations.append(
+                GateCalibration(str(gate), list(qubits), list(params), qobj_instructions))
+    return QasmExperimentCalibrations(gates=calibrations), pulse_library
+
+
+def _extract_common_calibrations(
+        experiments: List[QasmQobjExperiment]
+) -> Tuple[List[QasmQobjExperiment], Optional[QasmExperimentCalibrations]]:
+    """Given a list of ``QasmQobjExperiment``s, each of which may have calibrations in their
+    ``config``, collect common calibrations into a global ``QasmExperimentCalibrations``
+    and delete them from their local experiments.
+
+    Args:
+        experiments: The list of Qasm experiments that are being assembled into one qobj
+
+    Returns:
+        The input experiments with modified calibrations, and common calibrations, if there
+        are any
+    """
+    def index_calibrations() -> Dict[int, List[Tuple[int, GateCalibration]]]:
+        """Map each calibration to all experiments that contain it."""
+        exp_indices = defaultdict(list)
+        for exp_idx, exp in enumerate(experiments):
+            for gate_cal in exp.config.calibrations.gates:
+                # They must be keyed on the hash or identical cals will be indexed separately
+                exp_indices[hash(gate_cal)].append((exp_idx, gate_cal))
+        return exp_indices
+
+    def collect_common_calibrations() -> List[GateCalibration]:
+        """If a gate calibration appears in all experiments, collect it."""
+        common_calibrations = []
+        for _, exps_w_cal in exp_indices.items():
+            if len(exps_w_cal) == len(experiments):
+                _, gate_cal = exps_w_cal[0]
+                common_calibrations.append(gate_cal)
+        return common_calibrations
+
+    def remove_common_gate_calibrations() -> None:
+        """For calibrations that appear in all experiments, remove them from the individual
+        experiment's ``config.calibrations``."""
+        for _, exps_w_cal in exp_indices.items():
+            if len(exps_w_cal) == len(experiments):
+                for exp_idx, gate_cal in exps_w_cal:
+                    experiments[exp_idx].config.calibrations.gates.remove(gate_cal)
+
+    if not (experiments and all(hasattr(exp.config, 'calibrations') for exp in experiments)):
+        # No common calibrations
+        return experiments, None
+
+    exp_indices = index_calibrations()
+    common_calibrations = collect_common_calibrations()
+    remove_common_gate_calibrations()
```

Comment:
I see that you are trying to follow a pattern which splits up the work into smaller methods, but the explicit use of side effects on captured variables in the outer scope of `remove_common_gate_calibrations()` cuts pretty hard against any additional readability by breaking it out.

Reply:
Agree -- I'll update